pkg/services/oauthtoken/oauth_token.go

```diff
@@ -93,30 +101,54 @@ func (o *Service) GetCurrentOAuthToken(ctx context.Context, usr identity.Request
 
 	ctxLogger = ctxLogger.New("userID", userID)
 
-	authInfo, ok, _ := o.HasOAuthEntry(ctx, usr)
-	if !ok {
+	if !strings.HasPrefix(usr.GetAuthenticatedBy(), "oauth_") {
+		ctxLogger.Warn("The specified user's auth provider is not oauth",
+			"authmodule", usr.GetAuthenticatedBy())
 		return nil
 	}
 
-	if err := checkOAuthRefreshToken(authInfo); err != nil {
-		if errors.Is(err, ErrNoRefreshTokenFound) {
-			return buildOAuthTokenFromAuthInfo(authInfo)
+	var persistedToken *oauth2.Token
+	if o.features.IsEnabledGlobally(featuremgmt.FlagImprovedExternalSessionHandling) {
+		externalSession, err := o.sessionService.GetExternalSession(ctx, sessionToken.ExternalSessionId)
+		if err != nil {
+			if errors.Is(err, auth.ErrExternalSessionNotFound) {
+				return nil
+			}
+			ctxLogger.Error("Failed to fetch external session", "error", err)
+			return nil
 		}
 
-		return nil
-	}
+		persistedToken = buildOAuthTokenFromExternalSession(externalSession)
+
+		if persistedToken.RefreshToken == "" {
+			return persistedToken
+		}
```

Comment:
Does it make sense to check if the `persistedToken` contains a valid access token or this is covered by the `buildOAuthTokenFromExternalSession`?

Reply:
> Does it make sense to check if the persistedToken contains a valid access token or this is covered by the buildOAuthTokenFromExternalSession?

It shouldn't happen, so I don't think we should add this here. This particular check is for making sure that the `GetCurrentOAuthToken` function does not try to refresh a token without having a stored refresh token. Callers should validate whether the returned `*oauth2.Token` is not nil, invalid or expired.